code/API/FreshdeskClientFactory.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\Handler\CurlHandler;
+use GuzzleHttp\HandlerStack;
+use GuzzleHttp\Middleware;
+use SilverStripe\Framework\Injector\Factory;
+
+class FreshdeskClientFactory implements Factory
+{
+    const maxRetries = 5;
+
+    public function create($service, array $params = [])
+    {
+        $handlerStack = HandlerStack::create(new CurlHandler());
+        // setup a retry functionality that will retry the request self::maxRetries for connection errors and 500 status code
+        // errors. This will by default have an exponential back-off (sleep) between retries calculated as pow(2, $retries - 1);
+        $handlerStack->push(Middleware::retry(
+            function ($retries, $request, $response, $exception) {
+                if ($retries > self::maxRetries) {
+                    return false;
+                }
+                if ($exception instanceof ConnectException) {
+                    return true;
+                }
+                if ($response && $response->getStatusCode() >= 500) {
+                    return true;
+                }
+
+                return false;
+            },
+            function ($retries) {
+                // back-off, aka sleep between retries in milliseconds
+                return (int) 2 ** ($retries - 1) * 10;
+            }
+        ));
+        $client = new Client([
+            'handler' => $handlerStack,
+            'timeout' => 5.0,
```

Comment:
Some questions here:

* If the customer is submitting a ticket, will a timeout cause them to loose their form's content? :-) MANTIS BACK FROM THE DEAD maybe?
* Will the error surface appropriately so that they know how to resubmit?
* Would it make sense to extend the timeout? If a ticket is being submitted, I think it's probably OK to wait for longer - unless we use this also for time-critical, synchronous calls that would block the UI.

Reply:
Question still stands: will content be lost if an Exception is thrown from FreshdeskAPI::call during makeTicket?

Reply:
I don't have a good answer to this. If the API call fails a big ugly exception is thrown, an error is logged and the ticket does not exist in FD. The submitted data will exist in the SS CMS though so we can recover that way if needed.